File: examples/Rust/pi_digits.rs
// Build: rustc -O pi_digits.rs -o pi_digits_rs
// Run:   ./pi_digits_rs -r 2 -d 20 -t 2 -p
use std::env;
use std::thread;
use std::time::Instant;

fn pi_digits(count: usize) -> String {
    let mut a = vec![2_i64; count * 10 / 3 + 1];
    let (mut previous, mut pending) = (0_u8, 0_usize);
    let mut out = String::new();
    for _ in 0..=count {
        let mut q = 0_i64;
        for i in (1..=a.len()).rev() {
            let x = 10 * a[i - 1] + q * i as i64;
            a[i - 1] = x % (2 * i as i64 - 1);
            q = x / (2 * i as i64 - 1);
        }
        a[0] = q % 10;
        q /= 10;
        if q == 9 { pending += 1; }
        else if q == 10 {
            out.push((b'0' + previous + 1) as char);
            out.extend(std::iter::repeat_n('0', pending));
            previous = 0; pending = 0;
        } else {
            out.push((b'0' + previous) as char);
            out.extend(std::iter::repeat_n('9', pending));
            previous = q as u8; pending = 0;
        }
    }
    out.push((b'0' + previous) as char);
    out.extend(std::iter::repeat_n('9', pending));
    out[1..=count].to_string()
}

fn main() {
    let (mut runs, mut count, mut threads, mut print) = (1_usize, 10_usize, 1_usize, false);
    let mut args = env::args().skip(1);
    while let Some(arg) = args.next() {
        if arg == "-p" || arg == "--print" { print = true; continue; }
        let value = args.next().unwrap_or_else(|| panic!("missing value for {}", arg));
        let parsed = value.parse::<usize>().expect("expected a positive integer");
        assert!(parsed > 0, "counts must be positive");
        match arg.as_str() {
            "-r" | "--n-runs" => runs = parsed,
            "-d" | "--n-digits" => count = parsed,
            "-t" | "--n-threads" => threads = parsed,
            _ => panic!("unknown option: {}", arg),
        }
    }
    for run in 1..=runs {
        let run_start = Instant::now();
        let workers: Vec<_> = (0..threads).map(|_| thread::spawn(move || {
            let start = Instant::now();
            let digits = pi_digits(count);
            (digits, start.elapsed().as_secs_f64())
        })).collect();
        let mut first = String::new();
        for (j, worker) in workers.into_iter().enumerate() {
            let (digits, seconds) = worker.join().expect("worker failed");
            if j == 0 { first = digits; }
            if threads > 1 { println!("Run {run:3}/{runs:3}, thread {:3}/{threads:3} completed in {seconds:.6} seconds.", j + 1); }
            else { println!("Run {run:3}/{runs:3} completed in {seconds:.6} seconds."); }
        }
        if threads > 1 { println!("== Run {run:3}/{runs:3} completed in {:.6} seconds.", run_start.elapsed().as_secs_f64()); }
        if run == runs && print { println!("{first}"); }
    }
}

File: examples/C++/pi_digits.cpp
// Build: c++ -g -O0 -std=c++11 -pthread pi_digits.cpp -o pi_digits_cpp
// Run:   ./pi_digits_cpp -r 2 -d 20 -t 2 -p
#include <chrono>
#include <iostream>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>
#include <iomanip>

using Clock = std::chrono::steady_clock;
struct Result { std::string digits; double seconds; };

std::string pi_digits(int count) {
    std::vector<int> a(count * 10 / 3 + 1, 2);
    std::string out;
    int previous = 0, pending = 0;
    for (int step = 0; step <= count; ++step) {
        long long q = 0;
        for (int i = static_cast<int>(a.size()); i >= 1; --i) {
            long long x = 10LL * a[i - 1] + q * i;
            a[i - 1] = static_cast<int>(x % (2 * i - 1));
            q = x / (2 * i - 1);
        }
        a[0] = static_cast<int>(q % 10);
        q /= 10;
        if (q == 9) ++pending;
        else if (q == 10) {
            out += char('0' + previous + 1);
            out.append(pending, '0');
            previous = pending = 0;
        } else {
            out += char('0' + previous);
            out.append(pending, '9');
            pending = 0; previous = static_cast<int>(q);
        }
    }
    out += char('0' + previous);
    out.append(pending, '9');
    return out.substr(1, count);
}

int main(int argc, char **argv) {
    int runs = 1, count = 10, threads = 1;
    bool print = false;
    try {
        for (int i = 1; i < argc; ++i) {
            std::string arg = argv[i];
            if (arg == "-p" || arg == "--print") print = true;
            else if (i + 1 < argc && (arg == "-r" || arg == "--n-runs")) runs = std::stoi(argv[++i]);
            else if (i + 1 < argc && (arg == "-d" || arg == "--n-digits")) count = std::stoi(argv[++i]);
            else if (i + 1 < argc && (arg == "-t" || arg == "--n-threads")) threads = std::stoi(argv[++i]);
            else throw std::invalid_argument("unknown or incomplete option: " + arg);
        }
        if (runs < 1 || count < 1 || threads < 1) throw std::invalid_argument("counts must be positive");
        for (int run = 1; run <= runs; ++run) {
            auto run_start = Clock::now();
            std::vector<Result> results(threads);
            std::vector<std::thread> workers;
            for (int j = 0; j < threads; ++j) workers.emplace_back([&, j] {
                auto start = Clock::now();
                results[j].digits = pi_digits(count);
                results[j].seconds = std::chrono::duration<double>(Clock::now() - start).count();
            });
            for (int j = 0; j < threads; ++j) {
                workers[j].join();
                std::cout << std::fixed << std::setprecision(6);
                if (threads > 1) std::cout << "Run " << std::setw(3) << run << '/' << std::setw(3) << runs << ", thread " << std::setw(3) << j+1 << '/' << std::setw(3) << threads;
                else std::cout << "Run " << std::setw(3) << run << '/' << std::setw(3) << runs;
                std::cout << " completed in " << results[j].seconds << " seconds.\n";
            }
            if (threads > 1) std::cout << "== Run " << std::setw(3) << run << '/' << std::setw(3) << runs << " completed in " << std::chrono::duration<double>(Clock::now() - run_start).count() << " seconds.\n";
            if (run == runs && print) std::cout << results[0].digits << '\n';
        }
    } catch (const std::exception &e) { std::cerr << e.what() << '\n'; return 2; }
}
